pkg/steps/template.go

```diff
@@ -421,17 +421,57 @@ func createOrRestartPod(podClient coreclientset.PodInterface, pod *coreapi.Pod)
 	return created, nil
 }
 
-func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
+func waitForTeardownToComplete(podClient coreclientset.PodInterface, name string, uid types.UID) error {
 	timeout := 300
 	for i := 0; i < timeout; i += 2 {
-		pod, err := podClient.Get(name, meta.GetOptions{})
-		if errors.IsNotFound(err) {
-			return nil
+		pod, err := checkPodExists(podClient, name, uid)
+		if err != nil {
+			return err
+		}
+		if !podTeardownCompleted(pod) {
+			log.Printf("Waiting for pod %s to complete teardown ... (%ds/%d)", pod.GetName(), i, timeout)
+			time.Sleep(2 * time.Second)
+		} else {
+			break
 		}
+	}
+	return nil
+}
+
+func checkPodExists(podClient coreclientset.PodInterface, name string, uid types.UID) (*coreapi.Pod, error) {
+	pod, err := podClient.Get(name, meta.GetOptions{})
+	if errors.IsNotFound(err) {
+		return nil, errors.NewGone("pod was removed")
```

Comment:
notfound is expected in the waitForTeardownToComplete -- this should not be an error?

Reply:
Added description to the function. In short it would return error if the pod doesn't exist or has a different UID. In `waitForTeardownToComplete` and `waitForPodDeletion` this error doesn't get propagated though